File: backend/app/providers/direct.py
from __future__ import annotations

from urllib.parse import urlparse

from app.config import Settings
from app.errors import ApiError, file_too_large, unsupported_format
from app.models import MediaFormat
from app.providers.base import DownloadHandle, MediaMetadata, SocialMediaProvider
from app.providers.extractor import MediaExtractor
from app.providers.http import SafeHttp

VIDEO_EXTENSIONS = {".mp4", ".webm", ".mov", ".m4v", ".mkv"}
ALLOWED_MIME = {
    "video/mp4": "mp4",
    "video/webm": "webm",
    "video/quicktime": "mov",
    "video/x-m4v": "m4v",
    "video/x-matroska": "mkv",
    "video/mpeg": "mpg",
    "application/octet-stream": None,
}


class DirectVideoProvider(SocialMediaProvider):
    id = "direct"
    display_name = "Direct URL"
    supports_metadata = True
    supports_download = True
    notes = "Public video files and pages that expose a downloadable video."

    def __init__(
        self,
        settings: Settings,
        http: SafeHttp,
        extractor: MediaExtractor | None = None,
    ) -> None:
        self._settings = settings
        self._http = http
        self._extractor = extractor

    def can_handle(self, url: str) -> bool:
        path = urlparse(url).path.lower()
        return any(path.endswith(ext) for ext in VIDEO_EXTENSIONS)

    async def analyze(self, url: str) -> MediaMetadata:
        if self.can_handle(url):
            try:
                return await self._analyze_direct(url)
            except ApiError as exc:
                if exc.code in {"file_too_large", "invalid_url"}:
                    raise
                if self._extractor is not None:
                    return await self._extractor.analyze(
                        url,
                        platform=self.id,
                        display_name=self.display_name,
                    )
                raise
        if self._extractor is None:
            raise unsupported_format()
        return await self._extractor.analyze(
            url,
            platform=self.id,
            display_name=self.display_name,
        )

    async def create_download(self, url: str, format_id: str) -> DownloadHandle:
        if self.can_handle(url):
            try:
                meta = await self._analyze_direct(url)
                fmt = meta.formats[0] if meta.formats else None
                return DownloadHandle(
                    source_url=url,
                    format_id=format_id or "original",
                    mime_type=f"video/{fmt.format}" if fmt else "video/mp4",
                    filesize=fmt.filesize if fmt else None,
                    file_name=f"{meta.title}.{fmt.format if fmt else 'mp4'}",
                    upstream_url=url,
                )
            except ApiError:
                pass
        return DownloadHandle(
            source_url=url,
            format_id=format_id or "auto",
            mime_type="video/mp4",
            prepare_locally=True,
        )

    async def _analyze_direct(self, url: str) -> MediaMetadata:
        title = _filename(url)
        guessed = "mp4"
        path = urlparse(url).path.lower()
        for ext in VIDEO_EXTENSIONS:
            if path.endswith(ext):
                guessed = ext.lstrip(".")
                break
        try:
            response = await self._http.head_or_get(url)
        except ApiError:
            return MediaMetadata(
                platform=self.id,
                title=title,
                source_url=url,
                formats=[
                    MediaFormat(id="original", quality="original", format=guessed)
                ],
                can_download=True,
            )
        if response.status_code >= 400:
            return MediaMetadata(
                platform=self.id,
                title=title,
                source_url=url,
                formats=[
                    MediaFormat(id="original", quality="original", format=guessed)
                ],
                can_download=True,
            )
        content_type = (response.headers.get("content-type") or "").split(";")[0].strip().lower()
        fmt = _format_from_mime(content_type, url)
        length = _content_length(response.headers)
        if length is not None and length > self._settings.max_download_bytes:
            raise file_too_large()
        return MediaMetadata(
            platform=self.id,
            title=title,
            source_url=url,
            formats=[
                MediaFormat(
                    id="original",
                    quality="original",
                    format=fmt,
                    filesize=length,
                )
            ],
            can_download=True,
        )


def _format_from_mime(content_type: str, url: str) -> str:
    path = urlparse(url).path.lower()
    for ext in VIDEO_EXTENSIONS:
        if path.endswith(ext):
            guessed = ext.lstrip(".")
            break
    else:
        guessed = "mp4"
    if not content_type:
        return guessed
    if content_type.startswith("text/html") or content_type.startswith("application/json"):
        raise unsupported_format()
    if content_type in ALLOWED_MIME:
        return ALLOWED_MIME[content_type] or guessed
    if content_type.startswith("video/"):
        return content_type.split("/", 1)[1]
    raise unsupported_format()


def _content_length(headers) -> int | None:
    value = headers.get("content-length") or headers.get("content-range")
    if not value:
        return None
    if "content-range" in {k.lower() for k in headers.keys()} and "/" in str(
        headers.get("content-range", "")
    ):
        total = str(headers.get("content-range", "")).rsplit("/", 1)[-1]
        try:
            return int(total)
        except ValueError:
            return None
    try:
        return int(value)
    except ValueError:
        return None


def _filename(url: str) -> str:
    name = urlparse(url).path.rsplit("/", 1)[-1] or "video"
    if "." in name:
        name = name.rsplit(".", 1)[0]
    return name or "video"

File: backend/internal/providers/direct.go
package providers

import (
	"context"
	"net/url"
	"strings"

	"github.com/aviation256444-boop/socialsave/backend/internal/config"
	"github.com/aviation256444-boop/socialsave/backend/internal/errs"
	"github.com/aviation256444-boop/socialsave/backend/internal/models"
)

var videoExtensions = map[string]struct{}{
	".mp4": {}, ".webm": {}, ".mov": {}, ".m4v": {}, ".mkv": {},
}

var allowedMIME = map[string]string{
	"video/mp4":                "mp4",
	"video/webm":               "webm",
	"video/quicktime":          "mov",
	"video/x-m4v":              "m4v",
	"video/x-matroska":         "mkv",
	"video/mpeg":               "mpg",
	"application/octet-stream": "",
}

// Direct serves public video files and, for other pages, the shared extractor.
type Direct struct {
	cfg    config.Config
	fetch  *Fetcher
	runner *Runner
}

func (d *Direct) ID() string             { return "direct" }
func (d *Direct) SupportsMetadata() bool { return true }
func (d *Direct) SupportsDownload() bool { return true }
func (d *Direct) Notes() string {
	return "Public video files and pages that expose a downloadable video."
}

func (d *Direct) CanHandle(rawURL string) bool {
	path := strings.ToLower(pathOf(rawURL))
	for ext := range videoExtensions {
		if strings.HasSuffix(path, ext) {
			return true
		}
	}
	return false
}

func (d *Direct) Analyze(ctx context.Context, rawURL string) (Metadata, error) {
	if d.CanHandle(rawURL) {
		meta, err := d.analyzeFile(ctx, rawURL)
		if err == nil {
			return meta, nil
		}
		if api, ok := err.(*errs.Error); ok && (api.Code == "file_too_large" || api.Code == "invalid_url") {
			return Metadata{}, err
		}
		if d.runner != nil {
			return d.fromExtractor(ctx, rawURL)
		}
		return Metadata{}, err
	}
	if d.runner == nil {
		return Metadata{}, errs.UnsupportedFormat()
	}
	return d.fromExtractor(ctx, rawURL)
}

func (d *Direct) CreateDownload(ctx context.Context, rawURL, formatID string) (Handle, error) {
	if d.CanHandle(rawURL) {
		meta, err := d.analyzeFile(ctx, rawURL)
		if err == nil {
			format := "mp4"
			var size *int64
			if len(meta.Formats) > 0 {
				format = meta.Formats[0].Format
				size = meta.Formats[0].Filesize
			}
			if formatID == "" {
				formatID = "original"
			}
			return Handle{
				SourceURL:   rawURL,
				FormatID:    formatID,
				MimeType:    mimeForExt(format),
				Filesize:    size,
				FileName:    safeTitle(meta.Title) + "." + format,
				UpstreamURL: rawURL,
			}, nil
		}
	}
	if formatID == "" {
		formatID = "auto"
	}
	return Handle{
		SourceURL:      rawURL,
		FormatID:       formatID,
		MimeType:       "video/mp4",
		FileName:       "video.mp4",
		PrepareLocally: true,
	}, nil
}

func (d *Direct) analyzeFile(ctx context.Context, rawURL string) (Metadata, error) {
	title := fileTitle(rawURL)
	guessed := extFromPath(rawURL)
	resp, err := d.fetch.HeadOrRange(ctx, rawURL)
	if err != nil {
		if api, ok := err.(*errs.Error); ok && api.Code == "invalid_url" {
			return Metadata{}, err
		}
		return directMeta(title, rawURL, guessed, nil), nil
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return directMeta(title, rawURL, guessed, nil), nil
	}
	contentType := strings.ToLower(strings.TrimSpace(strings.Split(resp.Header.Get("Content-Type"), ";")[0]))
	format, err := formatFromMIME(contentType, rawURL)
	if err != nil {
		return Metadata{}, err
	}
	length := contentLength(resp.Header)
	if length != nil && *length > d.cfg.MaxDownloadBytes {
		return Metadata{}, errs.FileTooLarge(d.cfg.MaxDownloadBytes)
	}
	return directMeta(title, rawURL, format, length), nil
}

func (d *Direct) fromExtractor(ctx context.Context, rawURL string) (Metadata, error) {
	info, err := d.runner.Extract(ctx, rawURL, d.cfg)
	if err != nil {
		return Metadata{}, err
	}
	title := strings.TrimSpace(info.Title)
	if title == "" {
		title = "Direct URL video"
	}
	return Metadata{
		Platform:    "web",
		Title:       title,
		SourceURL:   rawURL,
		Thumbnail:   thumbnailOf(info),
		Duration:    durationOf(info),
		Author:      authorOf(info),
		Formats:     BuildFormats(info),
		CanDownload: true,
	}, nil
}

func directMeta(title, rawURL, format string, size *int64) Metadata {
	return Metadata{
		Platform:  "direct",
		Title:     title,
		SourceURL: rawURL,
		Formats: []models.MediaFormat{{
			ID:       "original",
			Quality:  "original",
			Format:   format,
			Filesize: size,
			HasAudio: true,
			HasVideo: true,
		}},
		CanDownload: true,
	}
}

func formatFromMIME(contentType, rawURL string) (string, error) {
	guessed := extFromPath(rawURL)
	if contentType == "" {
		return guessed, nil
	}
	if strings.HasPrefix(contentType, "text/html") || strings.HasPrefix(contentType, "application/json") {
		return "", errs.UnsupportedFormat()
	}
	if format, ok := allowedMIME[contentType]; ok {
		if format == "" {
			return guessed, nil
		}
		return format, nil
	}
	if strings.HasPrefix(contentType, "video/") {
		return strings.TrimPrefix(contentType, "video/"), nil
	}
	return "", errs.UnsupportedFormat()
}

func extFromPath(rawURL string) string {
	path := strings.ToLower(pathOf(rawURL))
	for ext := range videoExtensions {
		if strings.HasSuffix(path, ext) {
			return strings.TrimPrefix(ext, ".")
		}
	}
	return "mp4"
}

func fileTitle(rawURL string) string {
	name := pathOf(rawURL)
	if i := strings.LastIndex(name, "/"); i >= 0 {
		name = name[i+1:]
	}
	if name == "" {
		return "video"
	}
	if i := strings.LastIndex(name, "."); i > 0 {
		name = name[:i]
	}
	if name == "" {
		return "video"
	}
	return name
}

func pathOf(rawURL string) string {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return ""
	}
	return parsed.Path
}
